### DISCUSSION TITLE
[11.x] Remove Some Default Provider

### DISCUSSION THREAD

**Discussion by jyxjjj** [laravel/framework#50774] [Q&A]
At 10.x we can just remove providers by comment the lines in app.php,
while 10.x also moved default providers to vendors. So that we can continue copy it out to app.php then remove something.
How can we do it in 11.x?

> Of course I know we can continue delete things and downgrade to the behavior as 10.x too, but "upgrade" is more "sacred".

> DON'T TELL "upgrade application structure is not recommend".

Just like #50500 , I don't think there will be no one creates a new laravel project.

Btw, if someone who can resolve the issue of #50500 is better, thanks a lot.

**Comment by tayllana** [comment]
In Laravel 11.x, the structure of the application has been updated, and certain configurations have been moved to different locations compared to Laravel 10.x. If you want to remove or modify default service providers in Laravel 11.x without directly editing the `config/app.php` file, you can use service provider loading configurations available in the `config/app.php` file.

Here's how you can achieve this:

1. **Create a new service provider**: You can create a new service provider that loads only the necessary service providers you want to include in your application.

   ```bash
   php artisan make:provider CustomServiceProvider
   ```

2. **Modify the `register()` method in your custom service provider**: In the `register()` method of your custom service provider (`app/Providers/CustomServiceProvider.php`), you can specify which service providers to load.

   ```php
   <?php
   
   namespace App\Providers;
   
   use Illuminate\Support\ServiceProvider;
   
   class CustomServiceProvider extends ServiceProvider
   {
       /**
        * Register any application services.
        *
        * @return void
        */
       public function register()
       {
           // Load only the necessary service providers
           $this->app->register(DesiredServiceProvider::class);
           // Add more service providers as needed
       }
   }
   ```

3. **Register your custom service provider**: You need to register your custom service provider in the `config/app.php` file under the `providers` array.

   ```php
   'providers' => [
       // Other service providers
       App\Providers\CustomServiceProvider::class,
   ],
   ```

4. **Remove or comment out the default service provider**: You can now safely remove or comment out the default service provider from the `config/app.php` file if you no longer need it.

By following these steps, you can customize the loading of service providers in Laravel 11.x without directly modifying the `config/app.php` file. This approach allows you to maintain the upgrade integrity of your application while achieving the desired customization.

**Reply by jyxjjj** [reply]
What are you doing? ChatGPT?

**Comment by puklipo** [comment]
`providers` is still works in Laravel11.

### Remove default ServiceProviders
```php
// config/app.php

use Illuminate\Support\ServiceProvider;

    'providers' => ServiceProvider::defaultProviders()->except([
        Illuminate\Foo\FooServiceProvider::class,
    ])->toArray(),
```

### Replace
```php
// config/app.php

use Illuminate\Support\ServiceProvider;

    'providers' => ServiceProvider::defaultProviders()->replace([
        Illuminate\Foo\FooServiceProvider::class => Bar\BarServiceProvider::class,
    ])->toArray(),
```

- framework config includes `providers` https://github.com/laravel/framework/blob/efa867e85b4abcaab8146f66bc676b7239507719/config/app.php#L166-L173
- project config and framework config are merged and used.
- Just add `providers` to your project config.

**Reply by jyxjjj** [reply]
So, this should also delete bootstarp/providers.app ?

**Comment by deepCyan** [comment]
I have the same question...
I wanna remove the session provider but...you know that

**Reply by jyxjjj** [reply]
Yeah, i am now fallback to what 10.x did.
And my purpose is to remove something like broadcast and Authentication

**Reply by deepCyan** [reply]
actually, I get way that how to remove default providers just now, but about session...I MUST remove the default middleware for web route..It's a new question....why Laravel 11 do this...

**Reply by deepCyan** [reply]
you can add the config item in config/app.php

```php
    'providers' => ServiceProvider::defaultProviders()
        ->except([SessionServiceProvider::class]) // remove this
        ->merge([
            // Package Service Providers...
        ])->merge([
            // Application Service Providers...
            // App\Providers\AppServiceProvider::class,
        ])->merge([
            // Added Service Providers (Do not remove this line)...
        ])->toArray(),
```

**Reply by jyxjjj** [reply]
Right, it is more compact, and wordy and nagging. i want clean things.
I don't know if there was someone do malicious submission to damage this project.
I had tried some very simple pr too, and tyler even didn't take a look then refused.
I don't know what they want.

**Reply by deepCyan** [reply]
hi , I also found how to remove the default middleware, we should edit the bootstrap/app.php

```php
// code...
withMiddleware(function (Middleware $middleware) {
    $middleware->removeFromGroup('web', [
            Illuminate\Session\Middleware\StartSession::class,
            Illuminate\View\Middleware\ShareErrorsFromSession::class,
            Illuminate\Foundation\Http\Middleware\ValidateCsrfToken::class,
     ]);
    // BTW, you can call the $middleware->remove() to remove global default middleware...
})

```

**Comment by wizzymore** [comment]
You are trying to use the new structure in some parts but not in the others. Keep your laravel 10 structure, it is completely fine and 100% supported, you don't need to change a thing regarding configs/providers etc.

**Comment by AlexWaha** [comment]
The only way I found to exclude service providers from the boot is to update composer.json + `composer dump-autoload`, for example. RedisServiceProvider:
```json
  "extra": {
        "laravel": {
            "dont-discover": [
              "Illuminate\\Redis\\RedisServiceProvider"
            ]
        }
    },

**Reply by wizzymore** [reply]
Don't do this, this sucks and some other programmer will come by and don't know why the heck a service provider is not loaded, this is not standard for the base Illuminate service providers.

If you are still on the laravel 10 structure just comment the service provider you don't need, otherwise use the method from here: https://github.com/laravel/framework/discussions/50774#discussioncomment-8921269

**Reply by AlexWaha** [reply]
It's just an option 😄 

Otherwise, for 11-12.x you can use **config/app.php**

And add code like this, to replace the default provider:
```
 'providers' => ServiceProvider::defaultProviders()
  ->replace([
      \Illuminate\Auth\Passwords\PasswordResetServiceProvider::class => CustomPasswordResetProvider::class,
  ])->toArray(),
```
OR exclude some providers:

```
'providers' => ServiceProvider::defaultProviders()
    ->except([
        \Illuminate\Auth\Passwords\PasswordResetServiceProvider::class,
    ])->toArray(),
```
        
The default providers list can be found here:
        
`vendor/laravel/framework/src/Illuminate/Support/DefaultProviders.php`

**Reply by jyxjjj** [reply]
> [discussioncomment-12797338](https://github.com/laravel/framework/discussions/50774#discussioncomment-12797338)

Yeah I am still using things like this in 12.x

**Comment by macropay-solutions** [comment]
Why is laravel community pushing for this new bootstrap/app.php way of bootstraping the application in comparison with the old laravel 10?

Luckily for now both work.

Will they block the old way in the future?

**Reply by wizzymore** [reply]
Because it's way easier to keep all that logic in one file compared to 3 files in random and completely separate locations in the project.

And if you ever feel like the file is too big, there is nothing that prevents you from splitting it up in multiple files in your on structure.

**Reply by jyxjjj** [reply]
Same here — I’m worried it might get blocked someday.
If that actually happens, only then I’d think split and merge shouldn’t be handled by the framework.

**Reply by macropay-solutions** [reply]
Easier is to leave it as it was:)
